Many new features and fixes

HydrogenC opened the pull request:
Summary: 
**feat: a flutter-based ui of the curriculum board tab**
add: expanding the top banner
feat: random shuffle of banner items for fairness
remove: the unusable UIS fast login feature
fix: incorrect color of the like and dislike icons

The curriculum board section isn't adapted to the rewritten cb backend (yet), so expect changes when upstream updates.

HydrogenC commented:
The name `ivanfei` in the early commits isn't properly linked to the github account `ivanfei-1`, how to fix that?

w568w commented:
Wow, it looks a really large PR!


> **feat: a flutter-based ui of the curriculum board tab**

cc @ivanfei-1 

> remove: the unusable UIS fast login feature

What does this change do? I thought the UIS fast login feature has been removed in <https://github.com/DanXi-Dev/DanXi/commit/1cce976bb021bc9d0cd6d7edb26012676c910655>.

> The name `ivanfei` in the early commits isn't properly linked to the github account `ivanfei-1`, how to fix that? 

It seems to link correctly:

![](https://github.com/DanXi-Dev/DanXi/assets/8389787/b74f9b1f-0163-401e-9447-7c20bb443091)

HydrogenC commented:
> What does this change do? I thought the UIS fast login feature has been removed in [1cce976](https://github.com/DanXi-Dev/DanXi/commit/1cce976bb021bc9d0cd6d7edb26012676c910655).

It was my mistake. The title on that page was still "UIS quick login" which somehow misled me. In my new commit I reverted my changes and changed the title to "Login with UIS email". 


> It seems to link correctly:

There are some commits that doesn't link. 
![微信图片_20231002192405](https://github.com/DanXi-Dev/DanXi/assets/33123710/79e0e39c-ec4c-4b9c-86d6-11c07d289cdb)

w568w commented on lib/page/danke/course_review_editor.dart:
```diff
@@ -0,0 +1,757 @@
+
+class CourseReviewEditorWidgetState extends State<CourseReviewEditorWidget> {
+  late CourseReviewEditorText review;
+  final GlobalKey<DropdownListWidgetState<Course>> _timeSelectorKey =
+      GlobalKey<DropdownListWidgetState<Course>>();
+  final GlobalKey<DropdownListWidgetState<String>> _teacherSelectorKey =
+      GlobalKey<DropdownListWidgetState<String>>();
```
If you want to sync states between sub-`Widget`s (e.g. A and B are both sub-widgets of C; when A changes, B should change too), you should read about how `Provider` works: 

Simply put, **C holds the co-state of A and B** in a listenable way (e.g. a class `CoState` extends `ChangeNotifier`) and **mounts A and B under a Provider** (e.g. `ChangeNotifierProvider`). Then **A and B can watch it in their own `build()`** functions with `context.watch<CoState>()` or `Provider.of<CoState>(context)`. If one modifies the value, the other widget will be notified (i.e. rebuild) at once.

The co-pair you are using currently (i.e. `ValueNotifier` & `ValueListenableBuilder`) is a primitive way to accomplish this, provided by Flutter framework. However, let us hug more powerful Provider! :rocket:

w568w commented on lib/page/danke/course_sorts.dart:
What is this file for?

HydrogenC replied:
Legacy code from the original course-page branch, I will ask ivanfei what these are originally written for.

w568w commented on lib/page/subpage_danke.dart:
```diff
@@ -36,114 +40,81 @@ class DankeSubPage extends PlatformSubpage<DankeSubPage> {
+                    AnimatedContainer(
+                      duration: const Duration(milliseconds: 200),
+                      // Golden ratio minus the height of the title bar
+                      height: idle ? constraints.maxHeight * 0.382 - 56 : 0,
```
What if `constraints.maxHeight * 0.382 < 56`?

HydrogenC replied:
I will try to find a way to fetch the height of the page excluding the top bar and the bottom bar, so that it won't have to minus a 56.

w568w commented on lib/repository/opentreehole/opentreehole_repository.dart:
```diff
@@ -216,7 +216,7 @@ class OpenTreeHoleRepository extends BaseRepositoryWithDio {
         JWToken.fromJsonWithVerification(response.data!);
   }
 
-  Map<String, String> get _tokenHeader {
```
I thought it is a private getter. Why change its visibility to public?

HydrogenC replied:
My idea is to keep one shared jwt token between both treehole and danke. Maybe I should move it outside this class to show that it's shared.

w568w commented on lib/widget/danke/course_review_widget.dart:
```diff
@@ -0,0 +1,376 @@
+/*
+ *     Copyright (C) 2021  DanXi-Dev
```
Change the date in this copyright to the current year.

w568w commented on lib/widget/danke/random_review_widgets.dart:
```diff
@@ -0,0 +1,176 @@
+import 'package:flutter/cupertino.dart';
+import 'package:flutter/material.dart';
+
+import '../opentreehole/treehole_widgets.dart';
```
Do not use relative path import.

HydrogenC replied:
Mistake :(

w568w commented on lib/widget/danke/course_widgets.dart:
```diff
@@ -0,0 +1,246 @@
+        ? Theme.of(context).cardTheme.color?.withOpacity(0.8)
+        : null;
+
+int translateScore(int score) {
```
What does this mean? Please add some comments.

HydrogenC replied:
Unused legacy code, removed.

w568w commented on lib/widget/danke/course_widgets.dart:
```diff
@@ -0,0 +1,246 @@
+      : super(key: key);
+
+  @override
+  Widget build(BuildContext context) => Flex(
+          direction: Axis.horizontal,
+          mainAxisSize: MainAxisSize.min,
+          children: [
+            RoundChip(
+              onTap: onTap,
+              label: text,
+              color: color,
+            ),
+          ]);
```
`Flex` with `direction: Axis.horizontal` equals to `Row`. A `Row` with only one child seems unnecessary.

HydrogenC replied:
Since the code is copied from the treehole tags, the problem of the unnecessary Flex may also be present there.

w568w commented on lib/widget/danke/course_widgets.dart:
```diff
@@ -0,0 +1,246 @@
+  final int defaultIndex;
+}
+
+class FilterListWidgetState<T> extends State<FilterListWidget<T>> {
```
Consider adding a `didUpdateWidget()` too?

w568w commented on lib/widget/danke/random_review_widgets.dart:
```diff
@@ -0,0 +1,176 @@
+                                      OTLeadingTag(
+                                        color: Colors.orange,
+                                        text:
+                                            "${review.course!.credit!.toStringAsFixed(1)} ${S.of(context).credits}",
```
Consider merging this part with that in `lib/page/danke/course_group_detail.dart`? 

(Disclaimers: I have not compared them seriously.)

w568w commented on lib/widget/libraries/future_widget.dart:
```diff
@@ -141,3 +143,13 @@ class _FutureWidgetState<T> extends State<FutureWidget<T>> {
   }
 }
+
+Widget errorCard<T>(AsyncSnapshot<T> snapshot, void Function() onTap) => Card(
+      child: InkWell(
+        onTap: onTap,
+        child: Padding(
+          padding: const EdgeInsets.all(16.0),
+          child: Text("${snapshot.error}\n${snapshot.stackTrace}", maxLines: 10,),
+        ),
+      ),
+    );
```
This hard-coded part does not look like something that should appear in `widger/libraries`. Do you really use it heavily, or why not just use [`ErrorPageWidget.buildWidget`](https://github.com/DanXi-Dev/DanXi/blob/main/lib/widget/libraries/error_page_widget.dart)?

HydrogenC replied:
I would use ErrorPageWidget instead.

w568w commented on lib/widget/libraries/paged_listview.dart:
```diff
@@ -476,6 +476,10 @@ class PagedListViewState<T> extends State<PagedListView<T>>
     }
   }
 
+  Future<void> scrollToTop(
+          [Duration duration = kDuration, Curve curve = kCurve]) =>
+      currentController!.animateTo(0, duration: duration, curve: curve);
+
```
You does not need this method AFAICS. Here is a better way without any new method:  <https://github.com/DanXi-Dev/DanXi/blob/8d5261d7a6011a40d1c1956448c2263c1f0cef67/lib/page/subpage_treehole.dart#L593-L598> 

Just replace `PrimaryScrollController.of(context)` with your ScrollController. If you want to scroll the Treehole page to the top in it, `PrimaryScrollController.of(context)` is fine too.

w568w commented on lib/widget/opentreehole/auto_banner.dart:
```diff
@@ -23,22 +23,58 @@ import 'package:dan_xi/util/browser_util.dart';
+  List<BannerExtra?>? list;
+
+  // Hardcode the size for better performance and lower overhead
+  static const double bannerHeight = 40;
```
I am afraid that hardcoding the size will break the layout on devices with large font sizes (yes, I have tested and confirmed that). Why move from `SizedByChildBuilder`'s measurement?

HydrogenC replied:
My mistake, I thought that SizedByChildBuilder will also constrain the size of the builder element, but it actually doesn't.

w568w commented on lib/widget/opentreehole/auto_banner.dart:
```diff
@@ -57,47 +93,89 @@ class _AutoBannerState extends State<AutoBanner> {
+              _displayAll ? _buildAllList() : _buildSingleItem(),
+              SizedBox(
+                  height: 20,
+                  width: double.infinity,
```
You do not need to pass infinity on the unconstrained dimension. [Just omit it](https://api.flutter.dev/flutter/widgets/SizedBox/SizedBox.html).

w568w commented on lib/widget/opentreehole/treehole_widgets.dart:
```diff
@@ -942,15 +942,6 @@ class OTFloorWidgetBottomBarState extends State<OTFloorWidgetBottomBar> {
                           size: 12,
                         ),
                         text: S.of(context).modify,
-                        onTap: () async {
-                          Navigator.of(context).pop();
-                          if (await OTEditor.modifyReply(context, floor.hole_id,
-                              floor.floor_id, floor.content)) {
-                            if (!context.mounted) return;
-                            Noticing.showMaterialNotice(
-                                context, S.of(context).request_success);
-                          }
-                        },
```
Why remove these buttons' callbacks?

HydrogenC replied:
According to Boreaus, these callbacks are supposed to be handled in the onSelected event of the parent PopupMenuButton. These onTap are actually duplicate code that he forgotten to remove.

w568w reviewed (changes requested):
A preliminary review. Will test and report further problems if any.

HydrogenC commented on lib/model/danke/course_review.dart:
```diff
@@ -0,0 +1,135 @@
+  }
+
+  @override
+  int get hashCode => reviewId ?? timeCreated.hashCode;
```
Legacy mistake, fixed.

HydrogenC commented on lib/repository/danke/curriculum_board_repository.dart:
```diff
@@ -0,0 +1,126 @@
+  factory CurriculumBoardRepository.getInstance() => _instance;
+
+  // Return raw json string
+  Future<String?> getCourseGroups() async {
+    Response<String> response = await dio.get("$_BASE_URL/courses",
+        options: Options(headers: _tokenHeader));
+    return response.data;
```
All these will be removed or refractored when the golang version of the danke server which features searching gets ready.

HydrogenC commented on lib/page/danke/course_list_widget.dart:
```diff
@@ -0,0 +1,128 @@
+
+  XSharedPreferences? pref;
+
+  Future<List<CourseGroup>?> _fetchMegaList({bool forceRefetch = false}) async {
+    pref ??= await XSharedPreferences.getInstance();
+
+    // List is LARGE, avoid deserializing for a second time
+    if (_groups != null && !forceRefetch) {
+      return _groups;
+    }
+
+    String? coursesJson;
+    if (pref!.containsKey("course_groups") && !forceRefetch) {
+      coursesJson = pref!.getString("course_groups");
+    } else {
+      coursesJson =
+          await CurriculumBoardRepository.getInstance().getCourseGroups();
+      pref!.setString("course_groups", coursesJson);
+    }
+
+    List<dynamic> jsonArray = jsonDecode(coursesJson!);
+    return jsonArray.map((e) => CourseGroup.fromJson(e)).toList();
+  }
```
No longer needed with the new danke server.

HydrogenC commented on lib/page/danke/course_list_widget.dart:
```diff
@@ -0,0 +1,128 @@
+  List<CourseGroup>? _displayedGroups;
+  String? searchKeyword;
+
+  XSharedPreferences? pref;
```
Same as above.

HydrogenC commented:
@w568w All the requested changes so far have gotten resolved by now. Thanks for your detailed suggestion and instructions without which I couldn't have known how to fix these. Please review the added code since last review.

w568w commented on lib/page/danke/course_list_widget.dart:
```diff
@@ -0,0 +1,102 @@
+  @override
+  Widget build(BuildContext context) {
+    return Expanded(
+        child: PagedListView<CourseGroup>(
```
This is mostly due to poorly written documentation and comments, but I'll still point it out:

The behavior of `PagedListView` is very special: it (generally) does not directly accept the list data to be displayed, but **only rebuilds when the accepted parameters change by default**.

So, in your case, because **none of the parameters change when `searchKeyword` changes** (if you don't believe this, check parameters one by one yourself), **the list will also not reload itself when `searchKeyword` changes**, and you can verify this by yourself too.

---

The correct way to do this is to trigger a data refresh with the `PagedListViewController` in `didUpdateWidget()`. For example:

```dart
_listViewController.notifyUpdate(useInitialData: false, queueDataClear: false);
```

`useInitialData` indicates whether to discard the preloaded `initialData` previously passed to it in the parameter when refreshing; `queueDataClear` controls whether to keep the original contents of the list before refreshing (sometimes we want the user to be able to still see the old list items during refreshing, e.g., when the list of posts is being refreshed).

**Notes**: `notifyUpdate` will call `setState` inside the `PagedListView` too, but it is NOT redundant since `setState` is called on `PagedListViewState` rather than `CourseListWidgetState`.

w568w commented on lib/widget/danke/course_review_widget.dart:
```diff
@@ -0,0 +1,376 @@
+
+import 'package:dan_xi/model/danke/course_review.dart';
+
+class CourseReviewWidget extends StatelessWidget {
```
Haven't you decided where to show the title of reviews (i.e. `review.title`)?

w568w commented on lib/widget/danke/course_review_widget.dart:
```diff
@@ -0,0 +1,376 @@
+                        mainAxisAlignment: MainAxisAlignment.spaceBetween,
+                        children: [
+                          // use Expanded wrap the text to avoid overflow
+                          Expanded(
```
Same suggestion as below; scroll down.

w568w commented on lib/page/danke/course_group_detail.dart:
```diff
@@ -0,0 +1,469 @@
+
+  Widget _getListItems(BuildContext context,
+      ListProvider<CourseReview> dataProvider, int index, CourseReview review,
+      {bool isNested = false}) {
```
Remove unused params.

w568w commented on lib/page/danke/course_group_detail.dart:
```diff
@@ -0,0 +1,469 @@
+            }
+            // await refreshListView();
+            // // Set duration to 0 to execute [jumpTo] to the top.
+            // await _listViewController.scrollToIndex(0, const Duration());
+            // await scrollDownToFloor(e);
+          },
+          child: Text(S.of(context).modify),
+        ),
+
+      // Standard Operations
+      PlatformContextMenuItem(
+        menuContext: menuContext,
+        onPressed: () => smartNavigatorPush(context, "/text/detail",
+            arguments: {"text": e.filteredContent}),
+        child: Text(S.of(menuContext).free_select),
+      ),
+      PlatformContextMenuItem(
+          menuContext: menuContext,
+          child: Text(S.of(menuContext).copy),
+          onPressed: () async {
+            await FlutterClipboard.copy(renderText(e.filteredContent!, '', ''));
+            if (mounted) {
+              Noticing.showMaterialNotice(
+                  context, S.of(menuContext).copy_success);
+            }
+          }),
+      PlatformContextMenuItem(
+        menuContext: menuContext,
+        isDestructive: true,
+        onPressed: () async {
+          if (await OTEditor.reportPost(context, e.floor_id)) {
+            Noticing.showMaterialNotice(context, S.of(context).report_success);
+          }
+        },
+        child: Text(S.of(menuContext).report),
+      )
+    ];
+
+    return menu;
+  }
```
Remove the unused methods. Please ensure your changes are clean before committing with `dart fix --dry-run` & `dart fix`! Or you are just making the lives of both you and the reviewer heavier.

w568w commented on lib/page/danke/course_group_detail.dart:
```diff
@@ -0,0 +1,469 @@
+              future: _fetchCourseGroup(),
+              successBuilder: (context, snapshot) {
+                Future<List<CourseReview>?> allDataReceiver = _loadContent();
+                return PagedListView<CourseReview>(
+                  pagedController: _listViewController,
+                  withScrollbar: true,
+                  scrollController: PrimaryScrollController.of(context),
+                  // If we need to scroll to the end, we should prefetch all the data beforehand.
+                  // See also [prefetchAllFloors] in [TreeHoleSubpageState].
+                  allDataReceiver: allDataReceiver,
+                  builder: _getListItems,
+                  headBuilder: (ctx) => _buildHead(ctx),
+                  loadingBuilder: (BuildContext context) => Container(
+                    padding: const EdgeInsets.all(8),
+                    child: Center(child: PlatformCircularProgressIndicator()),
+                  ),
+                  endBuilder: (context) => Center(
+                    child: Padding(
+                      padding: const EdgeInsets.only(bottom: 16),
+                      child: Text(S.of(context).end_reached),
+                    ),
+                  ),
+                );
+              },
+              errorBuilder: (BuildContext context,
+                      AsyncSnapshot<CourseGroup?> snapshot) =>
+                  ErrorPageWidget.buildWidget(context, snapshot.error,
+                      stackTrace: snapshot.stackTrace,
+                      onTap: () => setState(() {})),
+              loadingBuilder:
+                  Center(child: PlatformCircularProgressIndicator())),
```
Add an `emptyBuilder` to ensure better experience in courses without any review.

w568w commented on lib/page/danke/course_list_widget.dart:
```diff
@@ -0,0 +1,102 @@
+        padding: const EdgeInsets.all(8),
+        child: Center(child: PlatformCircularProgressIndicator()),
+      ),
+      fatalErrorBuilder: (BuildContext context, dynamic snapshot) =>
```
Read its signature: `fatalErrorBuilder`'s params are `BuildContext context, dynamic error`, not `BuildContext context, dynamic snapshot`. `snapshot.error` will produce a `NoSuchMethodError` itself.

w568w commented on lib/page/subpage_danke.dart:
```diff
@@ -36,114 +42,92 @@ class DankeSubPage extends PlatformSubpage<DankeSubPage> {
-    _resetSubscription.cancel();
-  }
+  Future<CourseReview?> _loadRandomReview() async {
+    if (!context.read<FDUHoleProvider>().isUserInitialized) {
+      await OpenTreeHoleRepository.getInstance().initializeRepo();
+      settingsPageKey.currentState?.setState(() {});
```
`OpenTreeHoleRepository.getInstance().initializeRepo()` will throw a `NotLoginError` if user has not logged in FDUHole at all, and the error's message does not tell the user how to log in.

You'd better show a welcome page like `subpage_treehole` does when not logged in yet：

https://github.com/DanXi-Dev/DanXi/blob/8d5261d7a6011a40d1c1956448c2263c1f0cef67/lib/page/subpage_treehole.dart#L647-L657

Or, at least show a login button to user.

w568w reviewed (changes requested):
A lot to change. Haven't tested the appearance on Cupertino yet.

HydrogenC commented:
> Again, lots to change. Haven't tested the appearance on Cupertino yet.

According to KYLN, cupertino support and platform widgets are to be removed since there is already a swift version.

HydrogenC commented on lib/widget/danke/random_review_widgets.dart:
```diff
@@ -0,0 +1,175 @@
+                        padding: const EdgeInsets.symmetric(
+                            horizontal: 10, vertical: 2),
+                        child: Text(review.course!.code!,
+                            style: const TextStyle(color: Colors.white24))),
```
I missed that. Thanks for testing out.

HydrogenC commented on lib/widget/danke/random_review_widgets.dart:
```diff
@@ -0,0 +1,175 @@
+                              Padding(
+                                padding: const EdgeInsets.symmetric(
+                                    horizontal: 12, vertical: 2),
+                                child: Row(
```
Also legacy code. Removing it right away.

HydrogenC commented on lib/page/danke/course_list_widget.dart:
```diff
@@ -0,0 +1,102 @@
+  @override
+  void didUpdateWidget(covariant CourseListWidget oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    setState(() {
+      searchKeyword = widget.searchKeyword;
+    });
```
Got it.

HydrogenC commented on lib/widget/danke/random_review_widgets.dart:
```diff
@@ -0,0 +1,175 @@
+                Column(
+                  children: [
+                    // course name, department name, course code and credits
+                    Container(
```
Legacy code, refractoring right away.

HydrogenC commented on lib/widget/danke/random_review_widgets.dart:
```diff
@@ -0,0 +1,175 @@
+                    Row(
+                      mainAxisAlignment: MainAxisAlignment.spaceBetween,
+                      children: [
+                        // use Expanded wrap the text to avoid overflow
```
Legacy code. It can now wrap as expected after removing the redundant flexes.

w568w commented on lib/page/subpage_danke.dart:
```diff
@@ -36,114 +44,121 @@ class DankeSubPage extends PlatformSubpage<DankeSubPage> {
+                      ),
+                      CourseSearchBar(
+                        onSearch: (String text) {
+                          // Avoid unnecessary rebuilding
+                          if (text.isEmpty != idle) {
```
Are you sure that this `if` will avoid unnecessary rebuilding, rather than avoid ***any*** rebuilding except inputting the first character and deleting the last character? Now if I search "A" and change it to "B", searching will not be triggered at all!

HydrogenC replied:
I must be somehow unconscious when writing this code :(

w568w reviewed (changes requested):
And another thing I found necessary to be noticed: your `DropdownListWidget` **does not process long items very well** (check the course named 人工智能的伦理与治理 for an example):

![图片](https://github.com/DanXi-Dev/DanXi/assets/8389787/577242bd-d5d0-4cd0-b890-e675d801f91c)

According to [this answer](https://stackoverflow.com/a/55376107), you need to add the `isExpanded` property to `true` in `DropdownButton` (I think **this property is misnamed**. Read the documentation of this property and you will find **it not about expanding we thought at all**!)

Also, if you want a long label to be cut with `...`, add `overflow: TextOverflow.ellipsis` to your `Text` in `DropdownMenuItem` will do the trick.

w568w reviewed (approved):
There doesn't seem to be a significant problem anymore. Bravo for all the efforts so far! :rocket: 

@ivanfei-1 has told me that he has been particularly busy the last few days and may not be available to review this PR now, but will do so as soon as he is.

HydrogenC commented:
Thanks for your help.

ivanfei-1 commented:
Sorry for the late response. You two have been grilling me through email in the last couple of days ;p Excellent work on writing so many comments and reviews!

 Since w568w have carefully reviewed your code, I'll just check if there's any feature missing or modify some UI designs.

kooWZ commented:
In my phone, after clicking on a random view card, this record does not appear at the top of the redirected details page. Is it intended?
For example, the record appeared on home page is from User 257, which appears near the bottom in redirected detail page.
![c15f4d32b3c6037735b41bfd2217ff8](https://github.com/DanXi-Dev/DanXi/assets/60533182/e5a303e0-0827-4fdf-8062-0bc3936bb0b9)

kooWZ commented:
When creating a review, the star icons won't appear until you click on the blank area.
Also, the text exceeds screen's boundaries.
![3e3403f1c1559a201dd94d72f105054](https://github.com/DanXi-Dev/DanXi/assets/60533182/f09223ad-6cd6-404b-9f1e-3f460a1c7187)
![b99efd6b3aad2670d7f0b9da7efad53](https://github.com/DanXi-Dev/DanXi/assets/60533182/9620de61-85fb-4e93-bb3b-d672594721d2)

kooWZ commented:
https://github.com/HydrogenC/DanXi/blob/7665df79bf6c9315435da695d67272c2aa51b1b1/lib/page/danke/course_review_editor.dart#L144
I would recommend use a new phrase since the Chinese version is "发布新帖子"

HydrogenC commented:
> On my phone, after clicking on a random view card, this record does not appear at the top of the redirected details page. Is it intended? For example, the record showed on home page is from User 257, which appears near the bottom in redirected detail page. 
I'm using the scrollToItem function provided in the controller. Don't know exactly why it behaves like that.

HydrogenC commented:
> When creating a review, the star icons won't appear until you click on the blank area. Also, the text exceeds screen's boundaries. 

This is kind of weird since it the stars displays correctly on my PC and my PC Android Emulator. Further test needed.

ivanfei-1 commented:
> > When creating a review, the star icons won't appear until you click on the blank area. Also, the text exceeds screen's boundaries.
> 
> This is kind of weird since it the stars displays correctly on my PC and my PC Android Emulator. I don't know exactly why this is happening, further test needed. If this can be reproduced in the android emulator, please offer the detailed configurations.

This behavior also appears in the iOS simulators and the Cupertino version, maybe it is caused by the background color of the rating widget.

ivanfei-1 commented:
![20231011-212503](https://github.com/DanXi-Dev/DanXi/assets/67986259/e47a9470-eee6-4424-baa4-33e6acaec898)

There are also a few tiny issues about this page.
- The placeholder of the review filling widget is the same of the treehole's. But I think the problem that it isn't properly translated is because of some issues of the treehole widget back there.
- After clicking the fullscreen button on the top right corner and exiting the fullscreen mode, the curser will end up in the title box no matter where it was before.

ivanfei-1 commented:
![Screenshot 2023-10-11 at 21 38 08](https://github.com/DanXi-Dev/DanXi/assets/67986259/95530977-3e92-40cb-abe9-d615a46372b0)

After searching the courses, there isn't a way to go back to the front page or search for another course, and I'm stuck before I kill and reopen the app again.

HydrogenC commented:
> This behavior also appears in the iOS simulators and the Cupertino version, maybe it is caused by the background color of the rating widget.

You're right. Thanks for testing out. I forgot to replace the hardcoded icon color.

HydrogenC commented:
> ![Screenshot 2023-10-11 at 21 38 08](https://user-images.githubusercontent.com/67986259/274292546-95530977-3e92-40cb-abe9-d615a46372b0.png)
> 
> After searching the courses, there isn't a way to go back to the front page or search for another course, and I'm stuck before I kill and reopen the app again.

This is because that the cupertino appbar doesn't have a height. Some additional code will be required to fix this.

ivanfei-1 commented:
Also, please don't worry; these mistakes happen to all of us from time to time. Take your time to fix them and have a test. Overall, you have done a splendid job.
It would take a while before we can actually call it a final version for release because we do not have the procedure (or the test server) to test the pre-release version thoroughly just yet. If you feel the workload is becoming too much, each one of us is available to assist, so don't hesitate to ask for help.
I'll carefully check and make small adjustment on the UIs if you don't mind, as soon as I have the time to get my hands on the code ;P

HydrogenC commented:
> Also, please don't worry; these mistakes happen to all of us from time to time. Take your time to fix them and have a test. Overall, you have done a splendid job. It would take a while before we can actually call it a final version for release because we do not have the procedure (or the test server) to test the pre-release version thoroughly just yet. If you feel the workload is becoming too much, each one of us is available to assist, so don't hesitate to ask for help. I'll carefully check and make small adjustment on the UIs if you don't mind, as soon as I have the time to get my hands on the code ;P

Thanks for your encouragement. I don't really know how to design a UI with maximum usability, so feel free to adjust or even rewrite any of my designs.

w568w commented:
> I'm using the scrollToItem function provided in the controller. This isn't meant to behave like that, but I haven't figured out exactly why it does by now.

Haven't checked the code but I suppose it is due to forgetting considering the head builder's height when scrolling. I will inspect & fix that if my guess is correct.

> When creating a review, the star icons won't appear until you click on the blank area.

> The placeholder of the review filling widget is the same of the treehole's. 

> After clicking the fullscreen button on the top right corner and exiting the fullscreen mode, the curser will end up in the title box no matter where it was before.

Should be fixed in a95cef88f91130130f7d7bdba16de8ac59fae423.

> the text exceeds screen's boundaries.

Wrapped that in a `Wrap` could be a fast but dirty fix. I don't know, have a try. 

Anyway, after the between-space was adjusted, it is unlikely to overflow anymore (unless you have a *really* small screen!).

> After searching the courses, there isn't a way to go back to the front page or search for another course.

Should be fixed in 6c21d9c90d45b907705476f7318313db9e21b8e1.

---

Is there any other significant bug?

w568w commented:
@HydrogenC btw, which branch is yours (i.e. `course-page`) derived from, `course-page` or `main` in our repository?

w568w commented:
Kept existent and further reported issues here:

- [ ] The text exceeds screen's boundaries on very small screen
- [ ] Cannot modify own reviews

HydrogenC commented:
> @HydrogenC btw, which branch is yours (i.e. `course-page`) derived from, `course-page` or `main` in our repository?

It was derived from course-page, but I rebased it to main before I started adding my code.

HydrogenC commented:
> Keep existent and further reported issues here:
> 
> * [ ]  The text exceeds screen's boundaries on very small screen
> * [ ]  Cannot modify own reviews
> * [ ]  "课程风格" and "工作量" are wrongly rendered in each other's field

I will deal with these tonight. 
By the way, since the Swift version is already complete, why shall the flutter version continue supporting Cupertino?

w568w commented:
> By the way, since the Swift version is already complete, why shall the flutter version continue supporting Cupertino?

Swift version only supports iOS 17+ (approx; I'm not sure about the exact version. Ask @fsy2001 ), so the Flutter version will serve as a "Legacy" version for those who have not decided to jump into the latest iOS.

fsy2001 commented:
Swift version supports iOS 16.0+.

HydrogenC commented:
> Keep existent and further reported issues here:
> 
> * [ ]  The text exceeds screen's boundaries on very small screen
> * [ ]  Cannot modify own reviews
> * [ ]  "课程风格" and "工作量" are wrongly rendered

I believe the second and the third has already been resolved in [5f8f8b3](https://github.com/DanXi-Dev/DanXi/pull/296/commits/5f8f8b3a7a08a2fbecbfe26e217404d7f9e46ea0). As for the first, I really cannot come up with an idea how to make a usable layout on such small screens.

w568w commented:
@ivanfei-1 Is any further change required?

ivanfei-1 commented:
> @ivanfei-1 Is any further change required?

Efficient work ;)

A small one: the upvote and the downvote button have a weird, not correctly aligned gray background. I would suggest that you not use buttons with background.

And a **thicc** one: The badge feature is missing now, I didn't mention before because not only I haven't described the feature in details to @HydrogenC yet (you can grill me afterwards for under estimating the work efficiency of him) but also the previous UI design of the badges was quite vague: The red, yellow, blue dots on the right corner of each review card was originally design for this (yes, they are not traffic lights), after clicking them a tooltip would pop up to show the badges.

Since I will be *grilled* by my professors until this Wednesday midnight, if either of you have carefully review this pr, and as @w568w have mentioned, we should include this pr in 1.4.0, I would suggest that we hide the "traffic lights" and release without the badge feature, even if I think our colleagues at the Public Relation Department will be *pissed*🤣.

HydrogenC commented:
> A small one: the upvote and the downvote buttons have a weird, not correctly aligned gray background. I would suggest that you not use buttons with background.

I haven't really fully inspected the style of widgets under Cupertino. I'll try for a better design right away. 



> Since I will be _grilled_ by my professors until this Wednesday midnight, if either of you have carefully review this pr, and as @w568w have mentioned, we should include this pr in 1.4.0, I would suggest that we hide the "traffic lights" and release without the badge feature _for now_, even if I think our colleagues at the Public Relation Department will be _pissed_🤣.
I've heard about this before, but I suppose that this requires pretty much work since I am totally unfamiliar with the badge system. I also approve of not implementing it for now, and add it in a future hotfix release (i.e. 1.4.1).

w568w commented:
Finally merged. Looks like that we are approaching 1.4.0! :rocket:
